src/ansys/fluent/core/utils/file_transfer_service.py

```diff
@@ -0,0 +1,165 @@
+        except (AttributeError, KeyError):
+            pass
+        else:
+            from simple_upload_server.client import Client
+
+            self.file_service = Client(
+                token="token", url=upload_server.uri, headers=upload_server.headers
+            )
+
+    def _upload_download_helper(
+        self,
+        is_upload: bool,
+        file_name: Optional[str] = None,
+        remote_file_name: Optional[str] = None,
+        local_file_name: Optional[str] = None,
+    ):
+        """Uploads and downloads a file.
+
+        Parameters
+        ----------
+        is_upload : bool
+            True if we want to upload file, False otherwise.
+        file_name : str
+            file name
+        remote_file_name : str, optional
+            remote file name, by default None
+        local_file_name : str, optional
+            local file name, by default None
+        Raises
+        ------
+        FileNotFoundError
+            If the file does not exist.
+        """
+        if self.file_service:
+            if is_upload:
+                if os.path.isfile(file_name):
+                    expanded_file_path = os.path.expandvars(file_name)
+                    upload_file_name = remote_file_name or os.path.basename(
+                        expanded_file_path
+                    )
+                    self.file_service.upload_file(expanded_file_path, upload_file_name)
+                else:
+                    raise FileNotFoundError(f"{file_name} does not exist.")
+            else:
+                if self.file_service.file_exist(file_name):
+                    self.file_service.download_file(file_name, local_file_name)
+                else:
+                    raise FileNotFoundError("Remote file does not exist.")
+
+    def upload(self, file_name: str, remote_file_name: Optional[str] = None):
+        """Uploads a file on the server.
+
+        Parameters
+        ----------
+        file_name : str
+            file name
+        remote_file_name : str, optional
+            remote file name, by default None
+        Raises
+        ------
+        FileNotFoundError
+            If the file does not exist.
+        """
+        self._upload_download_helper(
+            is_upload=True, file_name=file_name, remote_file_name=remote_file_name
+        )
+
+    def download(self, file_name: str, local_file_name: Optional[str] = None):
+        """Downloads a file from the server.
+
+        Parameters
+        ----------
+        file_name : str
+            file name
+        local_file_name : str, optional
+            local file path, by default None
+
+        Raises
+        ------
+        FileNotFoundError
+            If the remote file does not exist.
+        """
+        self._upload_download_helper(
+            is_upload=False, file_name=file_name, local_file_name=local_file_name
+        )
+
+    def _pypim_upload(self, file_name: str, on_uploaded: Optional[Any] = None):
```

Comment:
I think the function name `"_pypim_upload"`, the argument name `"on_uploaded"` and the docstring are misleading: this function does not necessarily upload files using the PIM API, `on_uploaded` callback is executed even when the upload does not occur, and the docstring does not properly describe the function behavior or the arguments. As set up here, `on_uploaded` does not necessarily read a file, it is just a callback that will do something, usually read or write, with the file name, depending on how `_pypim_upload` is called. 

Quick name suggestions for you to consider would be, for example:
- `_pypim_upload` -> `_upload_if_configured`
- `on_uploaded` -> `callback` 

Also a quick question and suggestion: is using a callback a good idea here? The callback isn't very helpful as it can be replaced in less characters and lines than you have currently in the function callback definition and where the function is called, for example under `read_case` it could be simply:
```
self._pypim_upload(file_name)
self.file.read_case(file_name=file_name)
```
rather than what you suggest currently:
```
self._pypim_upload(
    file_name=file_name,
    on_uploaded=self.file.read_case,
)
```

Reply:
@seanpearsonuk That is a good point, in that case to reproduce the current behavior I believe it would be:
```
self._pypim_upload(file_name)
self.file.read_case(file_name=Path(file_name).name)
```
if using `pathlib`, else, as is currently done here
```
self._pypim_upload(file_name)
self.file.read_case(file_name=os.path.basename(file_name))
```

(Unless other file path/name processing is being done that I missed)

This is not a strong opinion from my side either, just a thought that occurred to me and I think could be considered, if it wasn't already

Reply:
`on_uploaded` was suggested by @seanpearsonuk. Earlier it was `api` and `api` may be `tui` or `file`.

Reply:
The class docstring already mentions about behavior but if it is not sufficient then may be we can update docstrings.

Reply:
is using a callback a good idea here? - I have added it considering point of reusability in other methods like `read_mesh` etc..

So for `read_mesh` we just need to change callback from `self.file.read_case` to `self.file.read_mesh`

Reply:
`Pathlib` approach looks good to me, we can integrate it current code but I just want to avoid separate line like `self.file.read_case(file_name=Path(file_name).name)` in each method.

I'm thinking we can do following change right ?

```
        if on_uploaded:
            on_uploaded(Path(file_name).name)
```